proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+## Design
 
-### Associating call tags
+### Tags
 
-When a module defines a function it can specify the call tag(s) that its associated `funcref` should accept.
-(By default, specifically the result of `dispatch.canon` on the signature is associated, making this backwards compatible.)
-So a module can explicitly specify no tags if it doesn't want the function to be implicitly indirectly callable, or some `new` tag if it only wants it to be indirectly callable with the module's own tag, or multiple tags if it wants to mix-and-match dispatching conventions.
-Of course, these associated call tags have to be compatible with the function's signature.
+In the Tags section, we add the following instructions for declaring/creating tags:
+* `call_tag.canon $functype` derives the canonical call tag of type `$functype` from the function type `$functype : [ti*] -> [to*]`, where all of `ti*` and `to*` are "concrete" (discussed [below](#asbstraction))
+* `call_tag.new $functype $func?` generates a new call tag of type `$functype = [ti*] -> [to*]` optionally with `$func` as its fall-back handler, where `$func : [ti* funcref] -> [to*]`
 
-### Calling with call tags
+Similarly, one can import and export call tags.
 
-`call_indirect` would then have a variant, say `call_funcref $ct` that specifies what call tag `$ct` to use.
-If the call tag `$ct` has associated signature type `[t1*] -> [t2*]`, then `call_funcref $ct` has type `[t1* funcref] -> [t2*]`.
-Thus the provided inputs and expected outputs are type-checked to match the signature associated with the call tag.
-The current `call_indirect` becomes just a shorthand for first getting the `funcref` from the appropriate table and then using `call_funcref` with the call tag resulting from `call_tag.canon` on the expected signature.
+### Functions
 
-The execution of `call_funcref $ct` on a given `funcref` for some function succeeds when the call tag `$ct` is one of the function's specified call tags.
-So if the function specified just one call tag, then a simply bitwise-equality check is done on the two tags at run time.
-If a match is found, then this *implies* that the arguments are valid inputs to the function and that the returned values are acceptable outputs from the function.
-However, by making call tags explicit, it also becomes clear that compatibility of inputs and outputs *does not imply* the call will succeed because the call tags might (intentionally) not match even if they have the same signature.
+When defining a `func` of type `[ti*] -> [to*]`, one can optionally specify `(call_tag $calltag*)`, where each `$calltag`'s type must be a supertype of with `[ti*] -> [to*]`.
+If so, the `funcref` returned by `func.ref` for this `func` handles exactly the call tags in `$calltag*`.
+
+`func_switch ($call_tag $func)* $func_switch` defines a new function, where the type of each `$call_tag` is a supertype of its corresponding `$func`'s type'.
```

Comment:
I'm confused -- where does this construct go? And why wouldn't calling it directly dump you in `$func_switch`?